Add tests and fix `insert_descriptor`

evanlinjin opened the pull request.

evanlinjin committed:
ref(chain): move `keychain::ChangeSet` into `txout_index.rs`

We plan to record `Descriptor` additions into persistence. Hence, we
need to add `Descriptor`s to the changeset. This depends on
`miniscript`. Moving this into `txout_index.rs` makes sense as this is
consistent with all the other files. The only reason why this wasn't
this way before, is because the changeset didn't need miniscript.

Co-Authored-By: Daniela Brozzoni <danielabrozzoni@protonmail.com>

danielabrozzoni committed:
ref(chain): Define test descriptors, use them...
...everywhere

danielabrozzoni committed:
keychain::ChangeSet includes the descriptor

- The KeychainTxOutIndex's internal SpkIterator now uses DescriptorId
  instead of K. The DescriptorId -> K translation is made at the
  KeychainTxOutIndex level.
- The keychain::Changeset is now a struct, which includes a map for last
  revealed indexes, and one for newly added keychains and their
  descriptor.

API changes in bdk:
- Wallet::keychains returns a `impl Iterator` instead of `BTreeMap`
- Wallet::load doesn't take descriptors anymore, since they're stored in
  the db
- Wallet::new_or_load checks if the loaded descriptor from db is the
  same as the provided one

API changes in bdk_chain:
- `ChangeSet` is now a struct, which includes a map for last revealed
  indexes, and one for keychains and descriptors.
- `KeychainTxOutIndex::inner` returns a `SpkIterator<(DescriptorId, u32)>`
- `KeychainTxOutIndex::outpoints` returns a `impl Iterator` instead of `&BTreeSet`
- `KeychainTxOutIndex::keychains` returns a `impl Iterator` instead of
  `&BTreeMap`
- `KeychainTxOutIndex::txouts` doesn't return a ExactSizeIterator
  anymore
- `KeychainTxOutIndex::unbounded_spk_iter` returns an `Option`
- `KeychainTxOutIndex::next_index` returns an `Option`
- `KeychainTxOutIndex::last_revealed_indices` returns a `BTreeMap`
  instead of `&BTreeMap`
- `KeychainTxOutIndex::reveal_to_target` returns an `Option`
- `KeychainTxOutIndex::reveal_next_spk` returns an `Option`
- `KeychainTxOutIndex::next_unused_spk` returns an `Option`
- `KeychainTxOutIndex::add_keychain` has been renamed to
  `KeychainTxOutIndex::insert_descriptor`, and now it returns a
  ChangeSet

danielabrozzoni commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -445,8 +445,12 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     ) -> super::ChangeSet<K> {
         let descriptor_id = descriptor.descriptor_id();
-        self.keychains_to_descriptors
+        let old_desc = self
+            .keychains_to_descriptors
             .insert(keychain.clone(), (descriptor_id, descriptor.clone()));
+        if let Some((old_desc_id, _)) = old_desc {
+            self.descriptor_ids_to_keychain.remove(&old_desc_id);
+        }
```
You're removing the old descriptor id if it got replaced, but you're not removing the old keychain id if the keychain got replaced. Is there a specific reason for this behavior? 

~~Also, is it really necessary to remove the old descriptor?~~ EDIT: yes it is!

danielabrozzoni commented on crates/chain/tests/test_keychain_txout_index.rs:
```diff
@@ -531,3 +531,67 @@ fn test_non_wildcard_derivations() {
     );
 }
+
+/// `::index_txout` should still index txouts with spks derived from descriptors without keychains.
```
Why is this desirable? Right now index_txout works only if you have the `keychain -> desc_id`, is this a problem?

evanlinjin replied:
For consistency. The keychain may be removed from a tracked descriptor and added later. You can see this scenario playing out in the test I introduced in this PR

evanlinjin replied:
Refer to my comment here: https://github.com/bitcoindevkit/bdk/pull/1203#discussion_r1522370903

notmandatory committed:
fix(wallet): add expected descriptors as signers after creating from wallet::ChangeSet

danielabrozzoni committed:
doc(bdk): Add instructions for manually inserting...
...secret keys in the wallet in Wallet::load

evanlinjin committed:
fix(chain): simplify `Append::append` impl for `keychain::ChangeSet`

We only need to loop though entries of `other`. The logic before was
wasteful because we were also looping though all entries of `self` even
if we do not need to modify the `self` entry.

evanlinjin committed:
test(chain): applying changesets one-by-one vs aggregate should be same

evanlinjin committed:
chore(chain): update test so clippy does not complain

evanlinjin committed:
chore(chain): move `use` in `indexed_tx_graph.rs` so clippy is happy

evanlinjin committed:
fix(chain): introduce keychain-variant-ranking to `KeychainTxOutIndex`

This fixes the bug where the result of applying individual changesets
may result in a different outcome than applying the aggregate changeset.

The nature of the changeset allows different keychain types to be
associated with the same descriptor. The previous design did not take
this into account.

Methods of `KeychainTxOutIndex` only returns one keychain type per
spk/txout/outpoint. This PR makes the keychain type returned to be
consistent by ranking keychain variants by `Ord`. The lowest keychain
variant (according to `Ord`) takes precedence.

evanlinjin committed:
chore(chain): simplify `keychains_to_descriptor_ids` field
